jsparagus/gen.py

```diff
@@ -1790,17 +1780,25 @@ def lookahead_lanes(self, state):
         def visit(aps):
             # Note, this suppose that we are not considering flags when
             # computing, as flag might prevent some lookahead investigations.
+            reduce_key = None
             first_reduce = next((e for e in aps.history[:-1] if not self.is_term_shifted(e.term)), None)
             if first_reduce:
-                reduce_key = (first_reduce, aps.shift[0].src, aps.history[-1].term)
-            has_seen_edge_after_reduce = first_reduce and reduce_key in seen_edge_after_reduce
+                is_reduced_term = len(aps.history) >= 2 \
+                    and not self.is_term_shifted(aps.history[-2].term)
```

Comment:
You mentioned this might be incorrect in some sense.

Reply:
It might be better to write this to be like the code that computes `first_reduce` above, but with the reverse iterator.

Reply:
I replaced this code by:

```python
            first_index, first_reduce = find_first_reduce(aps.history)
            last_index, last_reduce = find_last_reduce(aps.history)
            if first_index != last_index and first_reduce and last_reduce:
                if not isinstance(aps.history[-1].term, Action):
                    reduce_key = (first_reduce, aps.shift[0].src, last_reduce.term)
```
Where `find_first_reduce` and `find_last_reduce` are function which are walking `aps.hisotry` and return the index and the matching edge. Which I think makes this code simpler to follow.